ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,523 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.account;
+
+import com.github.ambry.commons.HelixPropertyStoreFactory;
+import com.github.ambry.commons.TopicListener;
+import com.github.ambry.config.HelixPropertyStoreConfig;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.helix.AccessOption;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.store.HelixPropertyStore;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * <p>
+ *   An implementation of {@link AccountService} that employs a {@link HelixPropertyStore} as its underlying storage.
+ *   It is internally configured with the store-relative path be {@link #FULL_ACCOUNT_METADATA_PATH} to fetch a full
+ *   set of {@link Account} metadata, and the same path to update {@link Account} metadata. The full {@link Account}
+ *   metadata will be cached locally, so serving an {@link Account} query will not incur any calls to the remote
+ *   {@code ZooKeeper} server, and the calls to a {@code HelixAccountService} is not blocking.
+ * </p>
+ * <p>
+ *   When a {@link HelixAccountService} starts up, it will automatically fetch a full set of {@link Account} metadata
+ *   and store them in the local cache. It also implements the {@link TopicListener} interface, and subscribes to a
+ *   {@link #ACCOUNT_METADATA_CHANGE_TOPIC}. Each time when receiving a {@link #FULL_ACCOUNT_METADATA_CHANGE_MESSAGE},
+ *   it will fetch the updated full {@link Account} metadata and refresh its local cache.
+ * </p>
+ * <p>
+ *   The full set of the {@link Account} metadata are stored in a single {@link ZNRecord}. The {@link ZNRecord}
+ *   contains {@link Account} metadata in a simple map in the following form:
+ * </p>
+ * <pre>
+ * {
+ *  {@link #ACCOUNT_METADATA_MAP_KEY} : accountMap &#60String, String&#62
+ * }
+ * </pre>
+ * <p>
+ * where {@code accountMap} is a {@link Map} from accountId String to {@link Account} JSON string.
+ * </p>
+ * <p>
+ *   Limited by {@link HelixPropertyStore}, the total size of {@link Account} data stored on a single {@link ZNRecord}
```

Comment:
It will be zk end error. An `update` operation in `AccountService` will fail because the return value from `helixStore.update` (which will actually call `zkClient.update`) will be `false`.